lib/src/https_rustls/configuration.rs

```diff
@@ -52,18 +54,71 @@ pub struct TlsApp {
 pub type HostName = String;
 pub type PathBegin = String;
 
+#[derive(thiserror::Error, Debug)]
+pub enum ListenerError {
+    #[error("failed to acquired the lock, {0}")]
+    LockError(String),
+    #[error("failed to handle certificate operation, got resolver error, {0}")]
+    ResolverError(GenericCertificateResolverError),
+}
+
 pub struct Listener {
     listener: Option<TcpListener>,
     address: SocketAddr,
     fronts: Router,
     answers: Rc<RefCell<HttpAnswers>>,
     config: HttpsListener,
     ssl_config: Arc<ServerConfig>,
-    resolver: Arc<CertificateResolverWrapper>,
+    resolver: Arc<MutexWrappedCertificateResolver>,
     pub token: Token,
     active: bool,
 }
 
+impl CertificateResolver for Listener {
+    type Error = ListenerError;
+
+    fn get_certificate(
+        &self,
+        fingerprint: &CertificateFingerprint,
+    ) -> Option<ParsedCertificateAndKey> {
+        let resolver = self
+            .resolver
+            .0
+            .lock()
+            .map_err(|err| ListenerError::LockError(err.to_string()))
+            .ok()?;
+
+        resolver.get_certificate(fingerprint)
+    }
+
+    fn add_certificate(
+        &mut self,
+        opts: &AddCertificate,
+    ) -> Result<CertificateFingerprint, Self::Error> {
+        let mut resolver = self
+            .resolver
+            .0
+            .lock()
+            .map_err(|err| ListenerError::LockError(err.to_string()))?;
+
+        resolver
+            .add_certificate(opts)
+            .map_err(|err| ListenerError::ResolverError(err))
+    }
+
+    fn remove_certificate(&mut self, opts: &RemoveCertificate) -> Result<(), Self::Error> {
+        let mut resolver = self
+            .resolver
+            .0
+            .lock()
+            .map_err(|err| ListenerError::LockError(err.to_string()))?;
+
+        resolver
+            .remove_certificate(opts)
+            .map_err(|err| ListenerError::ResolverError(err))
+    }
+}
+
```

Comment:
I don't really understand this whole certificate business, but this code looks beautifull overall.

Reply:
Thanks <3